hazelcast/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlan.java

```diff
@@ -504,10 +504,11 @@ private void initDag(InternalSerializationService jobSerializationService) {
     }
 
     private static Collection<? extends Processor> createProcessors(VertexDef vertexDef, int parallelism) {
-        final Collection<? extends Processor> processors = vertexDef.processorSupplier().get(parallelism);
-        if (processors.size() != parallelism) {
-            throw new JetException("ProcessorSupplier failed to return the requested number of processors." +
-                    " Requested: " + parallelism + ", returned: " + processors.size());
+        ProcessorSupplier processorSupplier = vertexDef.processorSupplier();
+        final Collection<? extends Processor> processors = processorSupplier.get(parallelism);
+        if (processorSupplier.checkNumberOfProcessors() && (processors.size() != parallelism)) {
```

Comment:
BTW> Respecting `localParallelism` and just returning no-op processor for the rest of the processors that won't get anything to consume will allow you to delete this code and PS' new method